app/code/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/AbstractEav.php

```diff
@@ -148,6 +148,21 @@ public function reindexAttribute($attributeId, $isIndexable = true)
     abstract protected function _prepareIndex($entityIds = null, $attributeId = null);
 
     /**
+     * Prepare product visibility attribute value index
+     *
+     * @param array $entityIds      the entity ids limitation
+     * @return $this
+     */
+    protected function _prepareVisibilityIndex($entityIds = null)
+    {
+        $attribute = $this->_eavConfig->getAttribute(\Magento\Catalog\Model\Product::ENTITY, 'visibility');
+        $this->_prepareIndex($entityIds, $attribute->getAttributeId());
```

Comment:
The additional call to _prepareIndex is necessary for this to work. Removing it results in the same incorrect visibility behavior.